core/src/main/java/org/apache/iceberg/avro/BuildAvroProjection.java

```diff
@@ -82,28 +82,48 @@ public Schema record(Schema record, List<String> names, Iterable<Schema.Field> s
+      Schema.Field avroField = updateMap.get(AvroSchemaUtil.makeCompatibleName(expectedField.name()));
 
       if (avroField != null) {
-        updatedFields.add(avroField);
-
+        if (expectedField.hasDefaultValue() && AvroSchemaUtil.isNonNullDefault(expectedField.getDefaultValue()) &&
+            (!avroField.hasDefaultValue() || avroField.defaultVal() == JsonProperties.NULL_VALUE)) {
+          Schema newFiledSchema = (expectedField.isOptional()) ?
+              AvroSchemaUtil.toOption(AvroSchemaUtil.convert(expectedField.type()), true) :
+              AvroSchemaUtil.convert(expectedField.type());
+          Schema.Field newField =
+              new Schema.Field(avroField.name(), newFiledSchema, avroField.doc(), expectedField.getDefaultValue());
+          newField.addProp(AvroSchemaUtil.FIELD_ID_PROP, expectedField.fieldId());
+          updatedFields.add(newField);
+        } else {
+          updatedFields.add(avroField);
+        }
       } else {
         Preconditions.checkArgument(
-            field.isOptional() || field.fieldId() == MetadataColumns.ROW_POSITION.fieldId(),
-            "Missing required field: %s", field.name());
-        // Create a field that will be defaulted to null. We assign a unique suffix to the field
-        // to make sure that even if records in the file have the field it is not projected.
-        Schema.Field newField = new Schema.Field(
-            field.name() + "_r" + field.fieldId(),
-            AvroSchemaUtil.toOption(AvroSchemaUtil.convert(field.type())), null, JsonProperties.NULL_VALUE);
-        newField.addProp(AvroSchemaUtil.FIELD_ID_PROP, field.fieldId());
+            expectedField.isOptional() || expectedField.fieldId() == MetadataColumns.ROW_POSITION.fieldId() ||
+            expectedField.hasDefaultValue(), "Missing required field: %s", expectedField.name());
+        // Create a field that will be defaulted to the expectedField's default value. If no default value,
+        // default to null and assign a unique suffix to the field to make sure that even if records in the
+        // file have the field it is not projected.
+        String newFieldName = expectedField.name();
+        Schema newFiledSchema;
+        Object defaultValue;
+        if (expectedField.hasDefaultValue() && AvroSchemaUtil.isNonNullDefault(expectedField.getDefaultValue())) {
+          newFiledSchema = AvroSchemaUtil.convert(expectedField.type());
+          defaultValue = expectedField.getDefaultValue();
+        } else {
+          newFieldName = newFieldName + "_r" + expectedField.fieldId();
+          newFiledSchema = AvroSchemaUtil.toOption(AvroSchemaUtil.convert(expectedField.type()));
+          defaultValue = JsonProperties.NULL_VALUE;
+        }
+        Schema.Field newField = new Schema.Field(newFieldName, newFiledSchema, null, defaultValue);
+        newField.addProp(AvroSchemaUtil.FIELD_ID_PROP, expectedField.fieldId());
```

Comment:
This branch seems to have common code with a the out `if` branch. Do you see room to combine and simplify?

Reply:
we have 4 cases (for 2 conditions), for three of them we create a new field and add to `updatedFileds`, but different fields. The only thing that can be done is to create a shallow 3-lines function with 5 argument. I prefer to keep as is